### PR TITLE
Removed RouteKeyHandling.CatchAll

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #3430 

@rynowak

### CODE REVIEW

File: samples/CustomRouteSample.Web/Startup.cs [hunk 1]
```diff
@@ -19,6 +19,10 @@ public void Configure(IApplicationBuilder app)
             app.UseMvc(routes =>
             {
                 routes.DefaultHandler = new LocalizedRoute(routes.DefaultHandler);
+                routes.MapRoute(
+                    "items",
+                    "api/Items/{country}/{action}",
+                    defaults: new { controller = "Items" });
```
Comment (rynowak) [inline samples/CustomRouteSample.Web/Startup.cs]: If these are APIs then use attribute routes

File: samples/CustomRouteSample.Web/Controllers/ItemsController.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Mvc;
+
+namespace CustomRouteSample.Web.Controllers
```
Comment (rynowak) [inline samples/CustomRouteSample.Web/Controllers/ItemsController.cs]: Why reuse an existing sample?

File: src/Microsoft.AspNet.Mvc.Core/Routing/ActionSelectionDecisionTree.cs [hunk 3]
```diff
@@ -41,69 +41,7 @@ public IReadOnlyList<ActionDescriptor> Select(IDictionary<string, object> routeV
...
-                    {
-                        filtered.Clear();
-                    }
-
-                    filtered.Add(action);
-                }
-            }
-
-            return filtered ?? results;
+            return results;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Routing/ActionSelectionDecisionTree.cs]: :heart_eyes_cat:

File: src/Microsoft.AspNet.Mvc.Core/Routing/ActionSelectionDecisionTree.cs [hunk 4]
```diff
@@ -157,11 +95,7 @@ public ActionDescriptorClassifier()
                     foreach (var constraint in item.RouteConstraints)
                     {
                         DecisionCriterionValue value;
-                        if (constraint.KeyHandling == RouteKeyHandling.CatchAll)
-                        {
-                            value = new DecisionCriterionValue(value: null, isCatchAll: true);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Routing/ActionSelectionDecisionTree.cs]: This code in routing can be simplified as well.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/CustomRouteSampleTest.cs [hunk 5]
```diff
@@ -54,5 +57,78 @@ public CustomRouteSampleTest(MvcTestFixture<CustomRouteSample.Web.Startup> fixtu
...
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Assert.Equal("Hello from everywhere", body);
+        }
+
+        private static LinkBuilder LinkFrom(string url)
+        {
+            return new LinkBuilder(url);
+        }
+
+        private class LinkBuilder
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/CustomRouteSampleTest.cs]: Unused?

**Comment by rynowak** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.FunctionalTests/CustomRouteSampleTest.cs (hunk 5 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/CustomRouteSampleTest.cs]: Yup. Will remove.

File: samples/CustomRouteSample.Web/Controllers/ItemsController.cs (hunk 2 above)
Comment (ajaybhargavb) [inline samples/CustomRouteSample.Web/Controllers/ItemsController.cs]: Isn't this too trivial of a feature to put it in a separate sample?

Comment (rynowak) [inline samples/CustomRouteSample.Web/Controllers/ItemsController.cs]: What I don't like is that they overlap. Both of theses are related to routing. It's simpler if they are separate.

File: src/Microsoft.AspNet.Mvc.Core/Routing/ActionSelectionDecisionTree.cs (hunk 4 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Routing/ActionSelectionDecisionTree.cs]: Done. https://github.com/aspnet/Routing/pull/259

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**